packages/block-editor/src/hooks/auto-inserting-blocks.js

```diff
@@ -0,0 +1,232 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { addFilter } from '@wordpress/hooks';
+import { Fragment } from '@wordpress/element';
+import { PanelBody, ToggleControl } from '@wordpress/components';
+import { createHigherOrderComponent } from '@wordpress/compose';
+import { createBlock, store as blocksStore } from '@wordpress/blocks';
+import { useDispatch, useSelect } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { InspectorControls } from '../components';
+import { store as blockEditorStore } from '../store';
+
+function AutoInsertingBlocksControl( props ) {
+	const { autoInsertedBlocksForCurrentBlock, groupedAutoInsertedBlocks } =
+		useSelect(
+			( select ) => {
+				const { getBlockTypes } = select( blocksStore );
+				const _autoInsertedBlocksForCurrentBlock =
+					getBlockTypes()?.filter(
+						( { autoInsert } ) =>
+							autoInsert && props.blockName in autoInsert
+					);
+
+				// Group by block namespace (i.e. prefix before the slash).
+				const _groupedAutoInsertedBlocks =
+					_autoInsertedBlocksForCurrentBlock?.reduce(
+						( groups, block ) => {
+							const [ namespace ] = block.name.split( '/' );
+							if ( ! groups[ namespace ] ) {
+								groups[ namespace ] = [];
+							}
+							groups[ namespace ].push( block );
+							return groups;
+						},
+						{}
+					);
```

Comment:
It's better to do filtering/grouping outside of the `mapSelect` to avoid accidental re-renders. We also added a warning to catch similar cases - #53666.

Reply:
Thank you, I just noticed the warning 😅 

Going to tackle as part of #54024 where I need to move things around anyway 😊

Reply:
I've tried to tackle this in 8b9713de4ed41922d634cb002a13af86730a0894 (part of https://github.com/WordPress/gutenberg/pull/54024).

However, even with that commit (and even when directly applied to `trunk`, with no further changes), I'm still getting the warning. I've tried to debug this, and the "changed" values that it seems to be complaining about are indeed identical 😬 I've noticed that this typically appears when `groupedAutoInsertedBlocks` is an empty object -- is it possible that the warning is checking reference equality?

```diff
diff --git a/packages/block-editor/src/hooks/auto-inserting-blocks.js b/packages/block-editor/src/hooks/auto-inserting-blocks.js
index ded15719b9..604982fb13 100644
--- a/packages/block-editor/src/hooks/auto-inserting-blocks.js
+++ b/packages/block-editor/src/hooks/auto-inserting-blocks.js
@@ -82,6 +82,15 @@ function AutoInsertingBlocksControl( props ) {
                                        {}
                                );
 
+                       console.log(
+                               'select',
+                               props.blockName,
+                               props.clientId,
+                               getBlockIndex( props.clientId ),
+                               getBlock( props.clientId )?.innerBlocks?.length,
+                               _rootClientId,
+                               _autoInsertedBlockClientIds
+                       );
                        return {
                                blockIndex: getBlockIndex( props.clientId ),
                                innerBlocksLength: getBlock( props.clientId )?.innerBlocks
```

![image](https://github.com/WordPress/gutenberg/assets/96308/98e952b8-8279-4159-b24f-4e57a6fdc12b)

Reply:
Looking at https://github.com/WordPress/gutenberg/pull/53666 and `isShallowEqual`, I guess that might be the case:

```
> let abc = { a: {} };
undefined
> let def = { a: {} };
undefined
> console.log( `Are abc and def equal? ${ eq( abc, def ) }` );
Are abc and def equal? false
```

How should we handle cases like these?

Reply:
It's probably complaining about `_autoInsertedBlockClientIds`. The `.reduce` will create a new array on each call.

Reply:
Yeah, that's the variable I meant. It seems to be happening when an empty object is returned for this. I think I have a fix, will push to https://github.com/WordPress/gutenberg/pull/53666.

Reply:
5d3eac0a38

Reply:
Thank you, @ockham!

I'll have a look later today or tomorrow morning. Can I apply testing instructions from this PR to #53666?

Reply:
Yes! Or alternatively use the ones from https://github.com/WordPress/gutenberg/pull/54024 (they're pretty much the same anyway, just in written form).

BTW I think I've gotten rid of _most_ of the warnings, but a few seem to persist (maybe one per auto-inserted block). It was easy enough to get rid of the ones that were comparing empty objects, but it's a bit harder to tackle the ones where `_autoInsertedBlockClientIds` is actually non-empty. LMK if you have any suggestions for those! 😄